shadow/read_shadow.go

```diff
@@ -0,0 +1,80 @@
+package shadow
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+)
+
+var ErrNoSuchUser = errors.New("shadow: user entry is not present in database")
+var ErrWrongPassword = errors.New("shadow: wrong password")
+
+// Read reads system shadow passwords database and returns all entires in it.
+func Read() ([]Entry, error) {
+	f, err := os.Open("/etc/shadow")
+	if err != nil {
+		return nil, err
+	}
+	scnr := bufio.NewScanner(f)
+
+	var res []Entry
+	for scnr.Scan() {
+		ent, err := parseEntry(scnr.Text())
+		if err != nil {
+			return res, err
+		}
+
+		res = append(res, *ent)
+	}
+	if err := scnr.Err(); err != nil {
+		return res, err
+	}
+	return res, nil
+}
+
+func parseEntry(line string) (*Entry, error) {
+	parts := strings.Split(line, ":")
+	if len(parts) != 9 {
+		return nil, errors.New("read: malformed entry")
```

Comment:
Can we use the package name instead in these errors? Otherwise just having "malformed entry" makes it hard to understand it's coming from the function reading `/etc/shadow`

Reply:
Somewhere deep in PAM code free() is called for response (what contradicts man page btw)
I get this when I try to use variable on stack:
```
==20147== Invalid free() / delete / delete[] / realloc()
==20147==    at 0x48389AB: free (vg_replace_malloc.c:530)
==20147==    by 0x4871D63: pam_vprompt (in /usr/lib/libpam.so.0.84.2)
==20147==    by 0x4871F3A: pam_prompt (in /usr/lib/libpam.so.0.84.2)
==20147==    by 0x486E367: ??? (in /usr/lib/libpam.so.0.84.2)
==20147==    by 0x5247810: pam_sm_authenticate (in /usr/lib/security/pam_unix.so)
==20147==    by 0x486CB5C: ??? (in /usr/lib/libpam.so.0.84.2)
==20147==    by 0x486C3AF: pam_authenticate (in /usr/lib/libpam.so.0.84.2)
==20147==    by 0x10934D: run (main.c:57)
==20147==    by 0x109416: main (main.c:78)
==20147==  Address 0x10c0b0 is 0 bytes inside data symbol "reply"
==20147== 
```

Reply:
Oh, right. PAM is terrible.